### PR TITLE
Export visits

### PR DESCRIPTION
**Pull request by PandaGrom**
# Description
[Export of Visits for Admins](https://trello.com/c/3OpqBz34/55-export-of-visits-for-admins)

## Type of change

- [ ] Bug fix
- [x] New feature 

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/assets/reports/visits_report_01-54-54 2022-02-07.csv [hunk 1]
```diff
@@ -0,0 +1,3 @@
+id,cabinet,date,notes,created_at,updated_at,user_id,patient_id,sick_leave_id,visit_id,diagnosis_ids
+1,407,2022-10-11 00:00:00 UTC,daaaa,2022-01-28 19:42:51 UTC,2022-01-28 20:37:17 UTC,23,1,3,,[2]
+3,511,2022-10-11 00:00:00 UTC,12312455463542,2022-01-30 14:27:57 UTC,2022-02-06 15:02:36 UTC,23,1,3,1,[2]
```
Comment (nieisnuje) [inline app/assets/reports/visits_report_01-54-54 2022-02-07.csv:3]: is it a template? if yes, why does it have some random data?

File: app/services/exports_manager/export_csv_visits.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    def filename
+      "visits_report_#{Time.zone.now.strftime('%T %F').tr(':', '-')}.csv"
+    end
+
+    def full_filename
+      @full_filename ||= Rails.root.join('app', 'assets', 'reports', filename)
+    end
+
+    def headers
+      %w[id cabinet date notes created_at updated_at user_id patient_id sick_leave_id visit_id diagnosis_ids]
```
Comment (nieisnuje) [inline app/services/exports_manager/export_csv_visits.rb]: move to constant
`HEADERS = %w[id cabinet date notes created_at updated_at user_id patient_id sick_leave_id visit_id diagnosis_ids].freeze`

File: app/services/exports_manager/export_csv_visits.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    end
+
+    private
+
+    def filename
+      "visits_report_#{Time.zone.now.strftime('%T %F').tr(':', '-')}.csv"
+    end
+
+    def full_filename
+      @full_filename ||= Rails.root.join('app', 'assets', 'reports', filename)
```
Comment (hannazspot) [inline app/services/exports_manager/export_csv_visits.rb:31]: 🙈

File: app/api/v1/admin/exports.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  module Admin
+    class Exports < API
+      helpers do
+        def visit
+          Visit.find(params[:id])
+        end
+      end
+
+      get :export_visits do
+        ExportVisits.perform_async(params)
```
Comment (hannazspot) [inline app/api/v1/admin/exports.rb:11]: are all params allowed??

File: app/services/concerns/with_apply_conditions.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        .pluck(:modifier)
+        .reduce(scope) { |s, modifier| apply_modifier(s, modifier) }
+
+      arel_conditions =
+        applicable_conditions.map do |condition|
+          if condition.is_a? Hash
+            condition[:condition]
+          else
+            condition
+          end
```
Comment (hannazspot) [inline app/services/concerns/with_apply_conditions.rb]: `condition.is_a?(Hash) ? condition[:condition] : condition`

File: app/services/exports_manager/export_csv_visits.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        visit.cabinet,
+        visit.date,
+        visit.notes,
+        visit.created_at,
+        visit.updated_at,
+        visit.user_id,
+        visit.patient_id,
+        visit.sick_leave_id,
+        visit.visit_id,
+        visit.diagnosis_ids
```
Comment (hannazspot) [inline app/services/exports_manager/export_csv_visits.rb]: ids don't provide ay valuable information for users

upload all info from related tables or don't add useless data to a report in general

File: app/services/visit_manager/list_visits.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      return if filters[:patient_id].blank?
+
+      table[:patient_id].eq(filters[:patient_id])
+    end
+
+    def with_sick_leave_condition
+      return unless filters[:with_sick_leave]
+
+      table[:sick_leave_id].not_eq(nil)
+    end
```
Comment (hannazspot) [inline app/services/visit_manager/list_visits.rb:60]: it's possible to do in SELECT section of a query as far as I got the logic, you have a lot of duplications

SELECT patient_id, start_at etc FROM ....

**Review by alhimick-prog — APPROVED** [review]

**Review by hannazspot — APPROVED** [review]

**Pull request merged**